Add an algorithm to get a storage bottle map given a storage key

yoshisatoyanagisawa opened the pull request:
This new algorithm, "obtain a local storage bottle map given a storage key", will be used by the ServiceWorker specification to allow obtaining a storage map directly with a storage key, rather than an environment.

Context: https://github.com/w3c/ServiceWorker/issues/1784.

<!--
Thank you for contributing to the Storage Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.

When you submit this PR, and each time you edit this comment (including checking a checkbox through the UI!), PR Preview will run and update it. As such make any edits in one go and only after PR Preview has run.

If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [x] At least two implementers are interested (and none opposed): n/a
   This is a clarification on the existing standard.
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   This is a clarification on the existing standard.
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed: n/a
   This is a clarification on the existing standard.
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed:n/a
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/storage_standard/185.html" title="Last updated on Aug 6, 2025, 2:09 AM UTC (6631090)">Preview</a> | <a href="https://whatpr.org/storage/185/b3c0fa9...6631090.html" title="Last updated on Aug 6, 2025, 2:09 AM UTC (6631090)">Diff</a>

yoshisatoyanagisawa committed:
Add an algorithm to get a storage bottle map given a storage key

This new algorithm, "obtain a local storage bottle map given a storage key",
will be used by the ServiceWorker specification to allow obtaining a storage
map directly with a storage key, rather than an environment.

Context: https://github.com/w3c/ServiceWorker/issues/1784.

yoshisatoyanagisawa committed:
minor fix.

yoshisatoyanagisawa committed:
Refactor: Consolidate storage bottle map acquisition logic

The `obtain a storage bottle map` (which works from an environment) and the
recently added `obtain a local storage bottle map given a storage key`
algorithms had significant logical overlap.

This change introduces a more fundamental algorithm, `obtain a storage bottle
map given a storage key`, which takes a `storage type` and `storage key` as
direct arguments and contains the core acquisition logic.

The other algorithms are refactored to call this new unified algorithm:
- `obtain a storage bottle map` is simplified to first resolve the environment
   to a storage key, then call the new function.
- `obtain a local storage bottle map given a storage key` becomes a simple
   wrapper that calls the new function with the type hardcoded to "local".

This refactoring removes code duplication and improves maintainability by
clarifying the separation of concerns.

yoshisatoyanagisawa committed:
Docs: Reorder algorithm definitions for clarity

Move the definition of the `obtain a storage bottle map given a storage key` algorithm before `obtain a storage bottle map`.

This ensures that algorithms are defined before they are referenced, improving the logical flow and readability of the specification.

yoshisatoyanagisawa commented:
@annevk Will you take a look?
This is for addressing https://github.com/w3c/ServiceWorker/issues/1784.

annevk commented on storage.bs:
```diff
@@ -438,10 +438,11 @@ standard and standards using this standard can access the contents.
    <a for="traversable navigable">storage shed</a>.
   </ol>
```
You removed _environment_ but here we are trying to get something from _environment_.

yoshisatoyanagisawa replied:
Good catch!
I have removed the dependency.  Please take another look.

yoshisatoyanagisawa committed:
Refactor: Decouple storage bottle map logic from environment

The "obtain a storage bottle map given a storage key" algorithm
previously had a dependency on `environment` to determine the
correct storage shed for session storage. This created an implicit
dependency that made the algorithm less modular.

This change refactors the algorithm to require the `storage shed`
to be passed in as an explicit argument. The responsibility of
determining the correct shed (either the user agent's for local
storage or the traversable navigable's for session storage) is
moved to the calling algorithms:
- `obtain a storage bottle map`
- `obtain a local storage bottle map given a storage key`

yoshisatoyanagisawa committed:
Refactor: Centralize storage shelf acquisition logic

Introduces a new internal algorithm, "get or create a storage shelf",
to consolidate the logic for retrieving or creating a storage shelf
from a given storage shed and key.

This change eliminates code duplication between:
- `obtain a storage shelf`
- `obtain a storage bottle map given a storage key`

Both algorithms are refactored to call this new centralized function,
improving maintainability and clarifying the underlying process.

yoshisatoyanagisawa committed:
Remove unnecessary space

yoshisatoyanagisawa commented:
Since we decided to go with https://github.com/w3c/ServiceWorker/pull/1786, let me abandon the PR.
We do not need to allow the cache to be looked up with a storage key.
